### PR TITLE
feat: use moving avg for sampling

### PR DESCRIPTION
**Pull request by s4kh**
Trying out moving average for sampling

## Changes

Please provide a brief description of the changes here.

## Merge requirement checklist

* [ ] Unit tests added/updated
* [ ] [`CHANGELOG.md`](https://github.com/grafana/grafana-opentelemetry-java) file updated for non-trivial changes
* [ ] Changes in public API reviewed (if applicable)

### CODE REVIEW

**Comment by zeitlinger** [comment]
The sampler should be registered in https://github.com/grafana/grafana-opentelemetry-java/blob/cb31391d38030cba9d8aa4871c4fad179e557be2/custom/src/main/java/com/grafana/extensions/GrafanaAutoConfigCustomizerProvider.java#L16

```java
  @Override
  public void customize(AutoConfigurationCustomizer autoConfiguration) {
    autoConfiguration
        .addPropertiesSupplier(GrafanaAutoConfigCustomizerProvider::getDefaultProperties)
        .addPropertiesCustomizer(TestedInstrumentationsCustomizer::customizeProperties)
        .addMeterProviderCustomizer(MetricsCustomizer::configure)
        .addSamplerCustomizer(sampler -> ...)
        .addResourceCustomizer(ResourceCustomizer::truncate);
  }
```

**Comment by s4kh** [comment]
yeah, I'll add those.

File: custom/src/main/java/com/grafana/extensions/processor/MovingAverageProcessor.java [hunk 1]
```diff
@@ -0,0 +1,76 @@
...
+      logger.log(
+          Level.INFO,
+          "avg {0} * threshold {1} = {2}, duration {3}",
+          new Object[] {avg, thresholdVal, avg * thresholdVal, duration});
+      // discard
+      if (duration < avg * thresholdVal) {
+        return;
+      }
+    }
+    logger.log(Level.INFO, "sending forward:{0}", new Object[] {span.getName()});
```
Comment (s4kh) [inline custom/src/main/java/com/grafana/extensions/processor/MovingAverageProcessor.java:69]: @zeitlinger do we need to call the `export` here?

**Comment by s4kh** [comment]
@zeitlinger When I configure process or exporter, the spans are not sent to the backend(tempo) - not able to visualize on grafana using https://github.com/grafana/docker-otel-lgtm/tree/main, I am able to receive it if I don't configure processor and exporter. Can you check if I am missing something?

**Comment by zeitlinger** [comment]
> Can you check if I am missing something?

I added a commit that should fix it

File: custom/src/main/java/com/grafana/extensions/processor/MovingAverageProcessor.java (hunk 1 above)
Comment (zeitlinger) [reply custom/src/main/java/com/grafana/extensions/processor/MovingAverageProcessor.java:69]: I added a commit that should fix it

File: custom/src/main/java/com/grafana/extensions/servertiming/ServerTimingHeaderReader.java [hunk 2]
```diff
@@ -51,7 +51,7 @@ public <REQUEST, RESPONSE> void consume(
                         GETTER);
             SpanContext spanContext = Span.fromContext(traceparent).getSpanContext();
             if (spanContext.getTraceFlags().isSampled()) {
-              DynamicSampler.setSampled(spanContext.getTraceId());
+              DynamicSampler.getInstance().setSampled(spanContext.getTraceId());
```
Comment (s4kh) [inline custom/src/main/java/com/grafana/extensions/servertiming/ServerTimingHeaderReader.java:54]: @zeitlinger can we get a `ConfigProperties` here?

File: custom/src/main/java/com/grafana/extensions/servertiming/ServerTimingHeaderCustomizer.java [hunk 3]
```diff
@@ -40,7 +40,7 @@ public <RESPONSE> void customize(
   static String toHeaderValue(Context context) {
     ReadWriteSpan span = (ReadWriteSpan) Span.fromContext(context);
     SpanContext spanContext = span.getSpanContext();
-    boolean sampled = DynamicSampler.evaluateSampled(span);
+    boolean sampled = DynamicSampler.getInstance().evaluateSampled(span);
```
Comment (s4kh) [inline custom/src/main/java/com/grafana/extensions/servertiming/ServerTimingHeaderCustomizer.java:43]: here as well.

File: custom/src/main/java/com/grafana/extensions/servertiming/ServerTimingHeaderReader.java (hunk 2 above)
Comment (zeitlinger) [reply custom/src/main/java/com/grafana/extensions/servertiming/ServerTimingHeaderReader.java:54]: you get the properties here: https://github.com/grafana/grafana-opentelemetry-java/blob/03f29b03d742361cae8f3bfdc4e48a67e76c43b3/custom/src/main/java/com/grafana/extensions/sampler/SamplingSpanProcessor.java#L37

then you can store it in the singleton

**Pull request merged**